### ISSUE TITLE
AMD iGPU (Ryzen 7000) and AMD dGPU (6950 XT) not working together 

### ISSUE THREAD

**Issue by tahmid-H** [terrymacdonald/DisplayMagician#251]
**Describe the bug**
I have 5 monitors. 2 on my desk, 3 on my simrig. One of the desk monitors are connected to the iGPU, and the other 4 monitors are directly into my dGPU. For some reason, when I set up multiple profiles where the iGPU monitor is disconnected through Windows' display settings, profiles where the iGPU monitor is enabled become inaccessible with the yellow triangle warning saying monitors in the profile are disconnected. Similarly, when the iGPU monitor is manually enabled through Windows' display settings, profiles where the iGPU monitor is disabled become inaccessible with the same warning.

When I completely unplugged the iGPU monitor from the motherboard, and setup new profiles with the 4 monitors connected to the dGPU, everything works flawlessly.

**To Reproduce**
Steps to reproduce the behaviour:
1. Enable all monitors. Make desk's ultrawide the main monitor. Save profile as "All (Desk Main)".
2. Disable all 3 sim monitors. Save profile as "Desk Only".
3. Enable all 3 sim monitors. Make center sim monitor the main monitor. Disable desk's ultrawide and iGPU monitor. Save profile as "Sim Only".
4. "All (Desk Main)" and "Desk Only" are inaccessible. 
5. Manually enable the iGPU monitor.
6. "All (Desk Main)" and "Desk Only" are accessible, but "Sim Only" inaccessible. 

**Expected behaviour**
All profiles should be accessible.

**Reporting Logs**
[DisplayMagician-Support-20230825-1649.zip](https://github.com/terrymacdonald/DisplayMagician/files/12443614/DisplayMagician-Support-20230825-1649.zip)

**Enviroment (please complete the following information):**
 - Windows Version: Windows 11 Pro
 - DisplayMagician Version 2.5.0.352
 - Video Card: MSI Radeon RX 6950 XT Gaming X Trio
 - CPU: AMD Ryzen 9 7900X
 - Motherboard: MSI MPG B650-I 
 - Number and make/model of monitors: 1 x Dell S3422DWG (dGPU DisplayPort). 3 x Acer XZ27U V (2x dGPU DisplayPort, 1x dGPU HDMI). 1 x HP 24w (1x iGPU HDMI)
 - Date of last video driver update: 6/26/2023 (Driver version 23.7.1)
 - Date of last windows update: 22H2 Build 22621.2215

**Comment by terrymacdonald** [comment]
Hi @tahmid-H ,

I've been doing some testing recently and I've found an issue with the logic for the displaymagician UI. The underlying display changing should still work, but the image shown in the UI and in the icon files created would have been wrong.

If not, this is a different issue and one we'd need to do some testing on.

Thanks
Terry

**Comment by terrymacdonald** [comment]
@tahmid-H ,

I'll be releasing a new test version of DisplayMagician in a week or so which should hopefully fix the issue I think you're experiencing. I'll drop a link in here when it is ready to go. If you'd be able to download and test to see if things are fixed that would be great.

Thanks
Terry

**Comment by terrymacdonald** [comment]
@tahmid-H  I've created a new test version of DisplayMagician that *may* fix the issue that you're experiencing, and I've attached it to this Github issue. Please download the attached zip file, extract the msi installer from the zip file, and then run the installer. It should upgrade your existing Display Profiles.

[DisplayMagicianSetup-v2.6.0.446.zip](https://github.com/terrymacdonald/DisplayMagician/files/14018624/DisplayMagicianSetup-v2.6.0.446.zip)

This new version of DisplayMagician has the ability to 'update' Display Profiles - effectively overwriting the display layout with the one that is currently being used. So if you start If you still start having issues with a Display Profile not working, simply:
- setup your display layout the way you want it
- then open DisplayMagician's Display Profiles Window
- select on the saved Display Profile you wish to update
- click the Update button
- say 'Yes' when asked if you really want to overwrite settings

Your DisplayMagician Display Profile will now be recreated using the current display settings, and they will overwrite the old settings. This will hopefully get rid of any display layout issues, and it doesn't require any changes to your Game Shortcuts. Nice!

There is a single known issue at the moment, which is that the NVIDIALibrary SetDisplayConfig function doesn't work for NVIDIA video cards. Luckily this issue shouldn't have much impact at all as the WindowsLibrary functions should do enough configuration of their own to ensure it all works. If you have some special NVIDIA settings that you use then this version of DisplayMagician *may* not apply them correctly. You will unfortunately just have to test if that is the case for your particular setup. 

I am currently troubleshooting the NVIDIALibrary SetDisplayConfig function error but it looks like it could be a bit of a pain to fix, so I didn't want to hold up the test release; I'd rather get the test version into your hands so I can see what other bugs are in there!

If this fixes your issue please reply here, and if you find another bug please log a new Github issue so I can track them separately.

Thanks
Terry

**Comment by terrymacdonald** [comment]
@tahmid-H  I did find some issues with the earlier v2.6.0.446 version. I've made a lot of revisions to how DisplayMagician works in v2.6.0.485 (attached) that should make it way more reliable. But it is highly likely you will need to 'update' your Display Profiles to make them work properly.

I'd recommend first of all applying a Display Profile, and then press the 'update' button. That will recreate the config for the Display Profile and should make it work properly. You'll need to do that to each of your saved Display Profiles. Then please test out changing between your Display Profiles to see if everything works as expected.

Here is the latest test version of DisplayMagician: 
[DisplayMagicianSetup-v2.6.0.485.zip](https://github.com/terrymacdonald/DisplayMagician/files/14075432/DisplayMagicianSetup-v2.6.0.485.zip)

Thanks for all your help testing.

Terry

**Comment by terrymacdonald** [comment]
@tahmid-H , I've added in more checks and logic to avoid crashes. Could you please try this latest version? 
[DisplayMagicianSetup-v2.6.0.508.zip](https://github.com/terrymacdonald/DisplayMagician/files/14135107/DisplayMagicianSetup-v2.6.0.508.zip)

I'm hoping this will work, but at the very least it should allow DisplayMagician to boot up enough for you to create a Support ZIP File. Let's see whether I've fixed it 🤞.

**Comment by terrymacdonald** [comment]
@tahmid-H The latest test version should be working properly, and hopefully fixes your issue. Could you please test it for me so I know that it fixes your issue? If it does then I'll be able to close this issue as the fix will be part of the next full release.

[DisplayMagicianSetup-V2.6.0.515-No-NvApi_SetDisplayConfig.zip](https://github.com/terrymacdonald/DisplayMagician/files/14200905/DisplayMagicianSetup-V2.6.0.515-No-NvApi_SetDisplayConfig.zip)

As I've not really had any feedback from you recently I'll give it a week and then close this issue if I have no further responses. We can always reopen this issue in the future if need be.

Thanks!

**Comment by tahmid-H** [comment]
@terrymacdonald Hey Terry. Sorry about not getting back to you - I haven't checked my mail in a while and didn't see the notification. Sadly, I don't have the AMD GPU anymore. I decided to get an Asus RTX 4080 due to the 5 ports it has, and it has been smooth sailing so far. Also, while I did have the AMD GPU and used the motherboard's port in conjunction, I had success using this software - https://sourceforge.net/projects/monitorswitcher/. I believe you've already fixed the issue, but if not, maybe that software could be a point of reference. Nonetheless, thank you for the effort you've put in to this!
